### PR TITLE
Create a QueryTermVector once and do the significance and connectedne…

### PR DESCRIPTION
**Pull request by baldersheim**
…ss lookups once.

@geirst and @havardpe PR

### CODE REVIEW

File: searchlib/src/vespa/searchlib/features/queryterm.h [hunk 1]
```diff
@@ -51,10 +61,20 @@ class QueryTermFactory {
      * @param lookupSignificance whether we should look up the significance for this term.
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/features/queryterm.h:61]: This parameter is removed, please update

File: searchlib/src/vespa/searchlib/features/queryterm.h [hunk 2]
```diff
@@ -51,10 +61,20 @@ class QueryTermFactory {
...
      * @param lookupConnectedness whether we should look up the connectedness this term has with the previous term.
      */
-    static QueryTerm create(const fef::IQueryEnvironment & env,
-                            uint32_t termIndex,
-                            bool lookupSignificance = true,
+    static QueryTerm create(const fef::IQueryEnvironment & env, uint32_t termIndex,
                             bool lookupConnectedness = false);
 };
 
+class QueryTermHelper {
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/features/queryterm.h:68]: Consider adding class comment.

**Review by geirst — APPROVED** [review]
👍

File: searchlib/src/vespa/searchlib/features/queryterm.cpp [hunk 3]
```diff
@@ -3,45 +3,66 @@
...
 
-QueryTerm
-QueryTermFactory::create(const IQueryEnvironment & env,
-                         uint32_t termIdx,
-                         bool lookupSignificance,
-                         bool lookupConnectedness)
+namespace {
+
+using QueryTermVectorWrapper = AnyWrapper<QueryTermVector>;
+const vespalib::string QUERY_TERMS_KEY("queryvectorhelper.queryterms");
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/features/queryterm.cpp:37]: 'queryvectorhelper' is a bit confusing, consider using 'querytermhelper.queryterms' instead

File: searchlib/src/vespa/searchlib/features/queryterm.cpp [hunk 4]
```diff
@@ -3,45 +3,66 @@
...
+    const Anything * obj = env.getObjectStore().get(QUERY_TERMS_KEY);
+    return (obj != nullptr) ? & static_cast<const QueryTermVectorWrapper *>(obj)->getValue() : nullptr;
+}
+
+QueryTermVector
+QueryTermHelper::createQueryTermvector(const IQueryEnvironment &env) {
+    QueryTermVector vector;
+    vector.reserve(env.getNumTerms());
+    for (size_t i(0); i < env.getNumTerms(); i++) {
+        vector.emplace_back(QueryTermFactory::create(env, i));
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/features/queryterm.cpp:63]: push_back is probably more appropriate for object copy

File: searchlib/src/vespa/searchlib/features/queryterm.cpp [hunk 5]
```diff
@@ -3,45 +3,66 @@
...
-    feature_t connectedness = 0;
-    if (lookupConnectedness) {
-        connectedness = search::features::util::lookupConnectedness(env, termIdx);
+    const Anything * obj = store.get(QUERY_TERMS_KEY);
+    if (obj == nullptr) {
+        store.add(QUERY_TERMS_KEY, std::make_unique<QueryTermVectorWrapper>(createQueryTermvector(env)));
+        obj = store.get(QUERY_TERMS_KEY);
     }
-    return QueryTerm(termData, significance, connectedness);
+    return static_cast<const QueryTermVectorWrapper *>(obj)->getValue();
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/features/queryterm.cpp:48]: consider adding a static convenience function to the AnyWrapper template that casts the Anything pointer to a const T reference.

**Review by havardpe — APPROVED** [review]

File: searchlib/src/vespa/searchlib/features/queryterm.h (hunk 1 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/features/queryterm.h]: Fixed

File: searchlib/src/vespa/searchlib/features/queryterm.h (hunk 2 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/features/queryterm.h:71]: Done

File: searchlib/src/vespa/searchlib/features/queryterm.cpp (hunk 3 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/features/queryterm.cpp]: fixed

File: searchlib/src/vespa/searchlib/features/queryterm.cpp (hunk 4 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/features/queryterm.cpp]: Fixed

File: searchlib/src/vespa/searchlib/features/queryterm.cpp (hunk 5 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/features/queryterm.cpp:48]: Fixed

**Pull request merged**